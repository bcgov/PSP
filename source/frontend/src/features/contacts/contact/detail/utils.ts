import { ContactInfoField } from 'features/contacts/interfaces';
import { Dictionary } from 'interfaces/Dictionary';
import { IContactAddress, IContactMethod } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export function getContactInfo(
  validTypes: Dictionary<string>,
  contactMethods?: IContactMethod[],
): ContactInfoField[] {
  if (contactMethods === undefined) {
    return [];
  }
  // Get only the valid types
  let filteredFields = contactMethods.reduce(
    (accumulator: ContactInfoField[], method: IContactMethod) => {
      if (Object.keys(validTypes).includes(method.contactMethodType.id)) {
        accumulator.push({
          info: method.value,
          label: validTypes[method.contactMethodType.id],
        });
      }
      return accumulator;
    },
    [],
  );

  // Sort according to the dictionary order
  return filteredFields.sort((a, b) => {
    return Object.values(validTypes).indexOf(a.label) - Object.values(validTypes).indexOf(b.label);
  });
}

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
