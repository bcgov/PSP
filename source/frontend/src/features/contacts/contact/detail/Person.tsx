import { FormSection } from 'components/common/form/styles';
import { AddressField } from 'features/contacts/interfaces';
import { IContactOrganization, IContactPerson } from 'interfaces/IContact';
import * as React from 'react';
import { Col } from 'react-bootstrap';
import { FaCircle, FaRegBuilding, FaRegUser } from 'react-icons/fa';
import { Link } from 'react-router-dom';

import * as Styled from '../../styles';
import { toAddressFields } from '../utils/contactUtils';
import ContactInfoSubForm from './ContactInfoSubForm';

export interface PersonViewProps {
  person: IContactPerson;
}

const PersonView: React.FunctionComponent<PersonViewProps> = ({ person }) => {
  let personAddresses: AddressField[];
  if (person.addresses === undefined) {
    personAddresses = [];
  } else {
    personAddresses = toAddressFields(person.addresses);
  }

  return (
    <>
      <FormSection key={'contact-person-' + person.id + '-names'} className="mb-4">
        <Styled.RowAligned>
          <Col>
            <Styled.H2 data-testid="contact-person-fullname">
              <FaRegUser size={20} className="mr-2" />
              {person.fullName}
            </Styled.H2>
          </Col>
          <Col md="auto" className="ml-auto">
            <Styled.StatusIndicators className={person.isDisabled ? 'inactive' : 'active'}>
              <FaCircle size={7} className="mr-2" />
              <span data-testid="contact-person-status">
                {person.isDisabled ? 'INACTIVE' : 'ACTIVE'}
              </span>
            </Styled.StatusIndicators>
          </Col>
        </Styled.RowAligned>
        <Styled.RowAligned>
          <Col md="auto">
            <strong>Preferred name:</strong>
          </Col>
          <Col md="auto">
            <span data-testid="contact-person-preferred">{person.preferredName}</span>
          </Col>
        </Styled.RowAligned>
      </FormSection>
      <FormSection key={'contact-person-' + person.id + '-organization'} className="mb-4">
        <Styled.RowAligned>
          <Col md="auto">
            <FaRegBuilding className="mr-2" />
            <strong>Organization(s):</strong>
          </Col>
          <Col md="auto">
            {person.organizations &&
              person.organizations.map((organization: IContactOrganization, index: number) => (
                <span key={'person-org-' + index}>
                  <Link
                    to={'/contact/O' + organization.id}
                    data-testid="contact-person-organization"
                  >
                    {organization.name}
                  </Link>
                  <br />
                </span>
              ))}
          </Col>
        </Styled.RowAligned>
      </FormSection>
      <FormSection key={'contact-person-' + person.id + '-contacts'} className="mb-4">
        <ContactInfoSubForm contactEntity={person} />
      </FormSection>
      <FormSection key={'contact-person-' + person.id + '-address'} className="mb-4">
        <Styled.H2Primary>Address</Styled.H2Primary>
        {personAddresses.map((field: AddressField, index: number) => (
          <Styled.RowAligned className="pb-3" key={'person-address-' + index}>
            <Col>
              <div className="pb-2">
                <strong>{field.label}</strong>
              </div>
              <span data-testid="contact-person-address">
                {field.streetAddress1 && <div>{field.streetAddress1} </div>}
                {field.streetAddress2 && <div>{field.streetAddress2} </div>}
                {field.streetAddress3 && <div>{field.streetAddress3} </div>}
                <div>{field.municipalityAndProvince} </div>
                {field.postal && <div>{field.postal} </div>}
                {field.country && <div>{field.country}</div>}
                {index + 1 !== personAddresses.length && <hr></hr>}
              </span>
            </Col>
          </Styled.RowAligned>
        ))}
      </FormSection>
      <FormSection key={'contact-person-' + person.id + '-comments'}>
        <Styled.RowAligned>
          <Col>
            <div>
              <strong>Comments:</strong>
            </div>
            <span data-testid="contact-person-comment">{person.comment}</span>
          </Col>
        </Styled.RowAligned>
      </FormSection>
    </>
  );
};

export default PersonView;
