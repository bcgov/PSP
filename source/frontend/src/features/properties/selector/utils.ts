import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { formatApiAddress, pidFormatter } from 'utils';

import { Api_Geometry, Api_Property } from './../../../models/api/Property';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
  ADDRESS = 'Address',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.longitude?.toFixed(6), property.latitude?.toFixed(6)]).join(', '),
    };
  } else if (property.address !== undefined) {
    return {
      label: NameSourceType.ADDRESS,
      value: property.address,
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getPrettyLatLng = (location?: Api_Geometry) =>
  compact([
    location?.coordinate?.x?.toFixed(6) ?? 0,
    location?.coordinate?.y?.toFixed(6) ?? 0,
  ]).join(', ');

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    return getApiPropertyName(property);
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getApiPropertyName = (property?: Api_Property): PropertyName => {
  if (property === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  let mapProperty: IMapProperty = {
    pin: property.pin?.toString(),
    pid: property.pid?.toString(),
    latitude: property.latitude,
    longitude: property.longitude,
    planNumber: property.planNumber,
    address: property.address !== undefined ? formatApiAddress(property.address) : undefined,
  };
  return getPropertyName(mapProperty);
};
