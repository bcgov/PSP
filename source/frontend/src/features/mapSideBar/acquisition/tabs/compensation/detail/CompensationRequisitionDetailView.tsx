import { useEffect, useState } from 'react';
import { Col, Row } from 'react-bootstrap';
import { FaExternalLinkAlt, FaMoneyCheck } from 'react-icons/fa';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import EditButton from '@/components/common/EditButton';
import { HeaderField } from '@/components/common/HeaderField/HeaderField';
import LoadingBackdrop from '@/components/common/LoadingBackdrop';
import { Section } from '@/components/common/Section/Section';
import { SectionField } from '@/components/common/Section/SectionField';
import { StyledSummarySection } from '@/components/common/Section/SectionStyles';
import { StyledAddButton } from '@/components/common/styles';
import { Claims, Roles } from '@/constants';
import useKeycloakWrapper from '@/hooks/useKeycloakWrapper';
import { Api_CompensationRequisition } from '@/models/api/CompensationRequisition';
import { Api_Organization } from '@/models/api/Organization';
import { Api_Person } from '@/models/api/Person';
import { Api_Product, Api_Project } from '@/models/api/Project';
import { formatMoney, prettyFormatDate } from '@/utils';
import { formatApiPersonNames } from '@/utils/personUtils';

import { DetailAcquisitionFileOwner } from '../../../models/DetailAcquisitionFileOwner';

export interface CompensationRequisitionDetailViewProps {
  compensation: Api_CompensationRequisition;
  compensationContactPerson: Api_Person | undefined;
  compensationContactOrganization: Api_Organization | undefined;
  acqFileProject?: Api_Project;
  acqFileProduct?: Api_Product | undefined;
  clientConstant: string;
  loading: boolean;
  setEditMode: (editMode: boolean) => void;
  onGenerate: (compensation: Api_CompensationRequisition) => void;
}

interface PayeeViewDetails {
  displayName: string;
  isGstApplicable: boolean;
  isPaymentInTrust: boolean;
  contactEnabled: boolean;
  contactString: string | null;
}

export const CompensationRequisitionDetailView: React.FunctionComponent<
  React.PropsWithChildren<CompensationRequisitionDetailViewProps>
> = ({
  compensation,
  compensationContactPerson,
  compensationContactOrganization,
  acqFileProject,
  acqFileProduct,
  clientConstant,
  loading,
  setEditMode,
  onGenerate,
}) => {
  const { hasClaim, hasRole } = useKeycloakWrapper();
  const [payeeDetails, setPayeeDetails] = useState<PayeeViewDetails | null>(null);

  const projectName =
    compensation?.alternateProject !== undefined
      ? compensation?.alternateProject?.code + ' - ' + compensation?.alternateProject?.description
      : '';

  useEffect(() => {
    if (!compensation) {
      setPayeeDetails(null);
      return;
    }

    let payeeDetail: PayeeViewDetails = {
      contactEnabled: false,
      isPaymentInTrust: compensation?.isPaymentInTrust || false,
      isGstApplicable: false,
      contactString: null,
      displayName: '',
    };

    if (!!compensation.acquisitionOwner) {
      const ownerDetail = DetailAcquisitionFileOwner.fromApi(compensation.acquisitionOwner);
      payeeDetail.displayName = ownerDetail.ownerName ?? '';
    } else if (compensation.interestHolderId) {
      if (compensationContactPerson) {
        payeeDetail.displayName = formatApiPersonNames(compensationContactPerson);
        payeeDetail.contactString = 'P' + compensationContactPerson?.id;
        payeeDetail.contactEnabled = true;
      } else if (compensationContactOrganization) {
        payeeDetail.displayName = compensationContactOrganization?.name ?? '';
        payeeDetail.contactString = 'O' + compensationContactOrganization.id;
        payeeDetail.contactEnabled = true;
      }
    } else if (compensation.acquisitionFileTeamId) {
      if (compensationContactPerson) {
        payeeDetail.displayName = formatApiPersonNames(compensationContactPerson);
        payeeDetail.contactString = 'P' + compensationContactPerson?.id;
        payeeDetail.contactEnabled = true;
      } else if (compensationContactOrganization) {
        payeeDetail.displayName = compensationContactOrganization?.name ?? '';
        payeeDetail.contactString = 'O' + compensationContactOrganization.id;
        payeeDetail.contactEnabled = true;
      }
    } else if (!!compensation.legacyPayee) {
      payeeDetail.displayName = `${compensation.legacyPayee}`;
    }

    var results =
      compensation.financials?.filter(el => {
        return el.isGstRequired === true;
      }) || [];

    payeeDetail.isGstApplicable = results.length > 0;

    setPayeeDetails(payeeDetail);
  }, [compensation, compensationContactOrganization, compensationContactPerson]);

  const compPretaxAmount = compensation?.financials
    .map(f => f.pretaxAmount ?? 0)
    .reduce((prev, next) => prev + next, 0);

  const compTaxAmount = compensation?.financials
    .map(f => f.taxAmount ?? 0)
    .reduce((prev, next) => prev + next, 0);

  const compTotalAmount = compensation?.financials
    .map(f => f.totalAmount ?? 0)
    .reduce((prev, next) => prev + next, 0);

  const userCanEditCompensationReq = (): boolean => {
    if (compensation.isDraft && hasClaim(Claims.COMPENSATION_REQUISITION_EDIT)) {
      return true;
    }

    if (!compensation.isDraft && hasRole(Roles.SYSTEM_ADMINISTRATOR)) {
      return true;
    }

    return false;
  };

  const editButtonBlock = (
    <EditButton
      title="Edit compensation requisition"
      onClick={() => {
        setEditMode(true);
      }}
    />
  );

  return (
    <StyledSummarySection>
      <LoadingBackdrop show={loading} parentScreen={true} />
      <Section>
        <StyledRow className="no-gutters">
          <Col xs="6">
            <HeaderField label="Client:" labelWidth="8" valueTestId="compensation-client">
              {clientConstant}
            </HeaderField>
            <HeaderField
              label="Requisition number:"
              labelWidth="8"
              valueTestId="compensation-number"
            >
              {compensation.isDraft ? 'Draft' : compensation.id}
            </HeaderField>
          </Col>
          <Col xs="6">
            <HeaderField
              label="Compensation amount:"
              labelWidth="8"
              contentWidth="4"
              valueTestId="header-pretax-amount"
            >
              <p className="mb-0 text-right">{formatMoney(compPretaxAmount ?? 0)}</p>
            </HeaderField>
            <HeaderField
              label="Applicable GST:"
              labelWidth="8"
              contentWidth="4"
              valueTestId="header-tax-amount"
            >
              <p className="mb-0 text-right">{formatMoney(compTaxAmount ?? 0)}</p>
            </HeaderField>
            <HeaderField
              label="Total cheque amount:"
              labelWidth="8"
              contentWidth="4"
              valueTestId="header-total-amount"
            >
              <p className="mb-0 text-right">{formatMoney(compTotalAmount ?? 0)}</p>
            </HeaderField>
          </Col>
        </StyledRow>
      </Section>

      <Section
        header={
          <FlexDiv>
            Requisition Details
            <RightFlexDiv>
              {setEditMode !== undefined && userCanEditCompensationReq() && editButtonBlock}

              <StyledAddButton
                onClick={() => {
                  onGenerate(compensation);
                }}
              >
                <FaMoneyCheck className="mr-2" />
                Generate H120
              </StyledAddButton>
            </RightFlexDiv>
          </FlexDiv>
        }
      >
        <SectionField label="Status" labelWidth="4" valueTestId="compensation-status">
          {compensation.isDraft ? 'Draft' : 'Final'}
        </SectionField>
        <SectionField label="Alternate project" labelWidth="4">
          {projectName}
        </SectionField>
        <SectionField label="Final date" labelWidth="4" valueTestId="compensation-finalized-date">
          {prettyFormatDate(compensation.finalizedDate)}
        </SectionField>
        <SectionField label="Agreement date" labelWidth="4">
          {prettyFormatDate(compensation.agreementDate)}
        </SectionField>
        <SectionField label="Expropriation notice served date" labelWidth="4">
          {prettyFormatDate(compensation.expropriationNoticeServedDate)}
        </SectionField>
        <SectionField label="Expropriation vesting date" labelWidth="4">
          {prettyFormatDate(compensation.expropriationVestingDate)}
        </SectionField>
        <SectionField
          label="Advanced payment served date"
          labelWidth="4"
          valueTestId="advanced-payment-served-date"
        >
          {prettyFormatDate(compensation.advancedPaymentServedDate)}
        </SectionField>
        <SectionField label="Special instructions" labelWidth={'12'}>
          {compensation.specialInstruction}
        </SectionField>
      </Section>

      <Section header="Financial Coding" isCollapsable initiallyExpanded>
        <SectionField label="Product" labelWidth="4">
          {acqFileProduct?.code ?? ''}
        </SectionField>
        <SectionField label="Business function" labelWidth="4">
          {acqFileProject?.businessFunctionCode?.code ?? ''}
        </SectionField>
        <SectionField label="Work activity" labelWidth="4">
          {acqFileProject?.workActivityCode?.code ?? ''}
        </SectionField>
        <SectionField label="Cost type" labelWidth="4">
          {acqFileProject?.costTypeCode?.code ?? ''}
        </SectionField>
        <SectionField label="Fiscal year" labelWidth="4">
          {compensation.fiscalYear ?? ''}
        </SectionField>
        <SectionField label="STOB" labelWidth="4">
          {compensation.yearlyFinancial?.code ?? ''} - {compensation.yearlyFinancial?.description}
        </SectionField>
        <SectionField label="Service line" labelWidth="4">
          {compensation.chartOfAccounts && (
            <label>
              {compensation.chartOfAccounts?.code} - {compensation.chartOfAccounts?.description}
            </label>
          )}
        </SectionField>
        <SectionField label="Responsibility centre" labelWidth="4">
          {compensation.responsibility && (
            <label>
              {compensation.responsibility?.code} - {compensation.responsibility?.description}
            </label>
          )}
        </SectionField>
      </Section>

      <Section header="Payment" isCollapsable initiallyExpanded>
        <SectionField label="Payee" labelWidth="4">
          <StyledPayeeDisplayName>
            {payeeDetails?.contactEnabled && payeeDetails?.contactString && (
              <StyledLink
                target="_blank"
                rel="noopener noreferrer"
                to={`/contact/${payeeDetails.contactString}`}
              >
                <span>{payeeDetails.displayName}</span>
                <FaExternalLinkAlt className="ml-2" size="1rem" />
              </StyledLink>
            )}

            {!payeeDetails?.contactEnabled && <label>{payeeDetails?.displayName ?? ''}</label>}
            {payeeDetails?.isPaymentInTrust && <label>, in trust</label>}
          </StyledPayeeDisplayName>
        </SectionField>
        <SectionField label="Amount (before tax)">
          {formatMoney(compPretaxAmount ?? 0)}
        </SectionField>
        <SectionField label="GST applicable?">
          {payeeDetails?.isGstApplicable ? 'Yes' : 'No'}
        </SectionField>
        {payeeDetails?.isGstApplicable && (
          <SectionField label="GST amount">{formatMoney(compTaxAmount ?? 0)}</SectionField>
        )}
        <SectionField label="Total amount">{formatMoney(compTotalAmount ?? 0)}</SectionField>
      </Section>

      <Section header="Financial Activities" isCollapsable initiallyExpanded>
        {compensation.financials.map((item, index) => (
          <>
            <StyledSubHeader>
              <label>Activity {index + 1}</label>
            </StyledSubHeader>
            <SectionField label="Code & Description" labelWidth="4">
              {item.financialActivityCode?.code} - {item.financialActivityCode?.description}
            </SectionField>
            <SectionField label="Amount (before tax)" labelWidth="4">
              {formatMoney(item.pretaxAmount ?? 0)}
            </SectionField>
            <SectionField label="GST applicable?" labelWidth="4">
              {item.isGstRequired ? 'Yes' : 'No'}
            </SectionField>
            <SectionField label="GST amount" labelWidth="4">
              {formatMoney(item.taxAmount ?? 0)}
            </SectionField>
            <SectionField label="Total amount" labelWidth="4">
              {formatMoney(item.totalAmount ?? 0)}
            </SectionField>
          </>
        ))}
      </Section>

      <Section>
        <SectionField label="Detailed remarks" labelWidth="12">
          {compensation.detailedRemarks}
        </SectionField>
      </Section>

      <Section>
        <StyledCompensationFooter>
          <Row>
            <Col className="pr-0 text-right">
              <label>Compensation amount:</label>
            </Col>
            <Col xs="3" className="pl-1 text-right">
              <span>{formatMoney(compPretaxAmount ?? 0)}</span>
            </Col>
          </Row>
          <Row>
            <Col className="pr-0 text-right">
              <label>Applicable GST:</label>
            </Col>
            <Col xs="3" className="pl-1 text-right">
              <span>{formatMoney(compTaxAmount ?? 0)}</span>
            </Col>
          </Row>
          <Row>
            <Col className="pr-0 text-right">
              <label>Total cheque amount:</label>
            </Col>
            <Col xs="3" className="pl-1 text-right">
              <span>{formatMoney(compTotalAmount ?? 0)}</span>
            </Col>
          </Row>
        </StyledCompensationFooter>
      </Section>
    </StyledSummarySection>
  );
};

export default CompensationRequisitionDetailView;

const StyledRow = styled(Row)`
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  font-weight: bold;
`;

const StyledLink = styled(Link)`
  display: flex;
  align-items: center;
`;

const FlexDiv = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 0.25rem;
`;

const RightFlexDiv = styled.div`
  display: flex;
  flex-direction: row-reverse;
`;

const StyledCompensationFooter = styled.div`
  font-size: 16px;
  font-weight: 600;
`;

const StyledPayeeDisplayName = styled.div`
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  text-align: left;
  height: 100%;
  overflow-y: auto;
  padding-bottom: 1rem;
`;

const StyledSubHeader = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  border-bottom: solid 0.1rem ${props => props.theme.css.primaryColor};
  margin-bottom: 2rem;

  label {
    color: ${props => props.theme.css.primaryColor};
    font-family: 'BCSans-Bold';
    font-size: 1.75rem;
    width: 100%;
    text-align: left;
  }

  button {
    margin-bottom: 1rem;
  }
`;
