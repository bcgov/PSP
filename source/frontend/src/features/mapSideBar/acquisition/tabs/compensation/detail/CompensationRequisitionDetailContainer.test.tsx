import { useApiContacts } from '@/hooks/pims-api/useApiContacts';
import {
  mockAcquisitionFileResponse,
  mockApiAcquisitionFileTeamOrganization,
  mockApiAcquisitionFileTeamPerson,
} from '@/mocks/acquisitionFiles.mock';
import { getMockApiDefaultCompensation } from '@/mocks/compensations.mock';
import { emptyApiInterestHolder } from '@/mocks/interestHolder.mock';
import { getMockOrganization } from '@/mocks/organization.mock';
import { Api_InterestHolder } from '@/models/api/InterestHolder';
import { Api_Person } from '@/models/api/Person';
import { render, RenderOptions, waitForEffects } from '@/utils/test-utils';

import {
  CompensationRequisitionDetailContainer,
  CompensationRequisitionDetailContainerProps,
} from './CompensationRequisitionDetailContainer';
import { CompensationRequisitionDetailViewProps } from './CompensationRequisitionDetailView';

jest.mock('@/hooks/pims-api/useApiContacts');
const getPersonConceptFn = jest.fn();
const getOrganizationConceptFn = jest.fn();
(useApiContacts as jest.Mock).mockImplementation(() => ({
  getPersonConcept: getPersonConceptFn,
  getOrganizationConcept: getOrganizationConceptFn,
}));

let viewProps: CompensationRequisitionDetailViewProps;

const CompensationViewComponent = (props: CompensationRequisitionDetailViewProps) => {
  viewProps = props;
  return <></>;
};

const setEditMode = jest.fn();

describe('Compensation Detail View container', () => {
  const setup = (
    renderOptions: RenderOptions & { props?: Partial<CompensationRequisitionDetailContainerProps> },
  ) => {
    const utils = render(
      <CompensationRequisitionDetailContainer
        View={CompensationViewComponent}
        {...renderOptions.props}
        loading={renderOptions.props?.loading ?? false}
        setEditMode={setEditMode}
        compensation={renderOptions.props?.compensation ?? getMockApiDefaultCompensation()}
        acquisitionFile={renderOptions.props?.acquisitionFile ?? mockAcquisitionFileResponse()}
        clientConstant={renderOptions.props?.clientConstant ?? '034'}
      />,
      {
        ...renderOptions,
        claims: renderOptions?.claims ?? [],
      },
    );

    return {
      ...utils,
    };
  };

  beforeEach(() => {
    getPersonConceptFn.mockResolvedValue({
      data: {
        id: 1,
        firstName: 'first',
        surname: 'last',
      } as Api_Person,
    });
    getOrganizationConceptFn.mockResolvedValue({
      data: getMockOrganization(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('makes request to get person concept for acquisition team payee', async () => {
    const teamPerson = mockApiAcquisitionFileTeamPerson();
    setup({
      props: {
        compensation: {
          ...getMockApiDefaultCompensation(),
          acquisitionFileTeamId: teamPerson.id ?? 1,
          acquisitionFileTeam: teamPerson,
        },
      },
    });

    await waitForEffects();
    expect(getPersonConceptFn).toHaveBeenCalled();
    expect(viewProps.compensationContactPerson).toStrictEqual({
      id: 1,
      firstName: 'first',
      surname: 'last',
    } as Api_Person);
  });

  it('makes request to get organization concept for acquisition team payee', async () => {
    const teamOrg = mockApiAcquisitionFileTeamOrganization();
    setup({
      props: {
        compensation: {
          ...getMockApiDefaultCompensation(),
          acquisitionFileTeamId: teamOrg.id ?? 1,
          acquisitionFileTeam: teamOrg,
        },
      },
    });

    await waitForEffects();
    expect(getOrganizationConceptFn).toHaveBeenCalled();
    expect(viewProps.compensationContactOrganization).toStrictEqual(getMockOrganization());
  });

  it('makes request to get person concept for interest holder payee', async () => {
    const ihPerson: Api_InterestHolder = {
      ...emptyApiInterestHolder,
      interestHolderId: 1,
      acquisitionFileId: 2,
      personId: 1,
      person: { id: 1, firstName: 'first', surname: 'last' },
    };
    setup({
      props: {
        compensation: {
          ...getMockApiDefaultCompensation(),
          interestHolderId: ihPerson.interestHolderId,
          interestHolder: ihPerson,
        },
      },
    });

    await waitForEffects();
    expect(getPersonConceptFn).toHaveBeenCalled();
    expect(viewProps.compensationContactPerson).toStrictEqual({
      id: 1,
      firstName: 'first',
      surname: 'last',
    } as Api_Person);
  });

  it('makes request to get organization concept for interest holder payee', async () => {
    const ihOrg: Api_InterestHolder = {
      ...emptyApiInterestHolder,
      interestHolderId: 1,
      acquisitionFileId: 2,
      organizationId: 100,
      organization: { id: 100, name: 'ABC Inc' },
    };

    setup({
      props: {
        compensation: {
          ...getMockApiDefaultCompensation(),
          interestHolderId: ihOrg.interestHolderId,
          interestHolder: ihOrg,
        },
      },
    });

    await waitForEffects();
    expect(getOrganizationConceptFn).toHaveBeenCalled();
    expect(viewProps.compensationContactOrganization).toStrictEqual(getMockOrganization());
  });
});
