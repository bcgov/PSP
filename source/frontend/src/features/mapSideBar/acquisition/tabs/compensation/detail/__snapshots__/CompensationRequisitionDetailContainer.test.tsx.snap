// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compensation Detail View container renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
</DocumentFragment>
`;
